File: android/joinery-android-dnsfilter/src/main/java/com/getjoinery/dnsfilter/DnsFilterModels.kt
package com.getjoinery.dnsfilter

import com.getjoinery.android.JsonValue

// Typed models over the `dns_filtering/` API surface
// (plugins/dns_filtering/docs/overview.md § API Surface). Every shape here is
// the JSON-clean export the server emits (ScrollDaddyHelper::exportDevice /
// exportBlock, account_summary, catalog) — no brand knowledge, so a second
// ScrollDaddy-style deployment (NetworkSentry) parses identically. Ported 1:1
// from the iOS JoineryDNSFilterKit models, which are unit-tested against the
// same fixtures.

// MARK: - devices

/**
 * One device from the `devices` action. [dohUrl] is what the standard-mode
 * VpnService forwards captured DNS queries to; [hardBlockHostnames] is what
 * strict mode's tunnel enforces at the connection level.
 */
data class DnsDevice(
    val deviceId: Int,
    val name: String,
    val deviceName: String,
    val deviceType: String,
    val timezone: String,
    val isActive: Boolean,
    val logQueries: Boolean,
    val filtersEditable: Boolean,
    val resolverUid: String,
    val dohUrl: String,
    val dotHostname: String,
    val hardBlockHostnames: List<String>,
    /** When the resolver last saw this device, or null. Proxied from the DNS
     *  server as an object (`{seen: ...}`); a bare string is tolerated too. */
    val lastSeen: String?,
    val blocks: List<DnsBlockSummary>,
) {
    /** The always-on block is the baseline policy every device carries. */
    val alwaysOnBlock: DnsBlockSummary? get() = blocks.firstOrNull { it.isAlwaysOn }
    val scheduledBlocks: List<DnsBlockSummary> get() = blocks.filter { !it.isAlwaysOn }

    companion object {
        fun from(json: JsonValue?): DnsDevice? {
            val j = json ?: return null
            val deviceId = j["device_id"]?.intValue ?: return null
            return DnsDevice(
                deviceId = deviceId,
                name = j["name"]?.stringValue ?: "",
                deviceName = j["device_name"]?.stringValue ?: "",
                deviceType = j["device_type"]?.stringValue ?: "",
                timezone = j["timezone"]?.stringValue ?: "UTC",
                isActive = j["is_active"]?.boolValue ?: true,
                logQueries = j["log_queries"]?.boolValue ?: false,
                filtersEditable = j["filters_editable"]?.boolValue ?: true,
                resolverUid = j["resolver_uid"]?.stringValue ?: "",
                dohUrl = j["doh_url"]?.stringValue ?: "",
                dotHostname = j["dot_hostname"]?.stringValue ?: "",
                hardBlockHostnames = (j["hard_block_hostnames"]?.arrayValue ?: emptyList())
                    .mapNotNull { it.stringValue },
                lastSeen = parseLastSeen(j["last_seen"]),
                blocks = (j["blocks"]?.arrayValue ?: emptyList())
                    .mapNotNull { DnsBlockSummary.from(it) },
            )
        }

        /** `last_seen` arrives as an object (`{seen: "..."}`) proxied from the
         *  DNS server, as a bare string, or null. Fold all three to the seen
         *  timestamp. An empty object (PHP `[]`) reads as null. */
        private fun parseLastSeen(value: JsonValue?): String? {
            if (value == null || value.isNull) return null
            value.get("seen")?.stringValue?.let { return it }
            return value.stringValue
        }
    }
}

/**
 * The `blocks` summary rows on a device (no rule contents — [ruleCount] only).
 * The full contents come from `scheduled_block_edit`.
 */
data class DnsBlockSummary(
    val blockId: Int,
    val name: String,
    val isAlwaysOn: Boolean,
    val isActive: Boolean,
    val activeNow: Boolean,
    val ruleCount: Int,
    val schedule: DnsSchedule,
) {
    companion object {
        fun from(json: JsonValue?): DnsBlockSummary? {
            val j = json ?: return null
            val blockId = j["block_id"]?.intValue ?: return null
            return DnsBlockSummary(
                blockId = blockId,
                name = j["name"]?.stringValue ?: "",
                isAlwaysOn = j["is_always_on"]?.boolValue ?: false,
                isActive = j["is_active"]?.boolValue ?: true,
                activeNow = j["active_now"]?.boolValue ?: false,
                ruleCount = j["rule_count"]?.intValue ?: 0,
                schedule = DnsSchedule.from(j["schedule"]),
            )
        }
    }
}

data class DnsSchedule(
    val start: String?,
    val end: String?,
    val days: List<Int>,
    val timezone: String?,
) {
    companion object {
        fun from(json: JsonValue?): DnsSchedule = DnsSchedule(
            start = json?.get("start")?.stringValue,
            end = json?.get("end")?.stringValue,
            days = (json?.get("days")?.arrayValue ?: emptyList()).mapNotNull { it.intValue },
            timezone = json?.get("timezone")?.stringValue,
        )
    }
}

// MARK: - block contents (scheduled_block_edit read)

/**
 * Full contents of one block: category filters, service toggles, and custom
 * domain rules. The always-on editor loads this to render its Block/Allow
 * state. [filters]/[services] map key -> action (0 block, 1 allow); an absent
 * key means Allow ("no row").
 */
data class DnsBlockContents(
    val blockId: Int,
    val name: String,
    val isAlwaysOn: Boolean,
    val schedule: DnsSchedule,
    val filters: Map<String, Int>,
    val services: Map<String, Int>,
    val rules: List<DnsDomainRule>,
) {
    companion object {
        fun from(data: JsonValue?): DnsBlockContents? {
            if (data == null) return null
            // scheduled_block_edit read wraps the block under `block`; be lenient
            // and accept a flat shape too.
            val block = data["block"] ?: data
            val blockId = block["block_id"]?.intValue ?: return null
            return DnsBlockContents(
                blockId = blockId,
                name = block["name"]?.stringValue ?: "",
                isAlwaysOn = block["is_always_on"]?.boolValue ?: false,
                schedule = DnsSchedule.from(block["schedule"]),
                filters = actionMap(block["filters"]),
                services = actionMap(block["services"]),
                rules = (block["rules"]?.arrayValue ?: emptyList()).mapNotNull { DnsDomainRule.from(it) },
            )
        }

        /** The block's filter/service rows arrive either as a list of
         *  `{key/filter_key, action}` objects or as a `{key: action}` map,
         *  depending on the action. An empty PHP assoc array serializes as `[]`,
         *  which [JsonValue.objectValue] reads as an empty object. Fold all
         *  forms into `{key: action}`. */
        private fun actionMap(value: JsonValue?): Map<String, Int> {
            if (value == null) return emptyMap()
            val out = LinkedHashMap<String, Int>()
            val rows = value.arrayValue
            if (rows != null) {
                for (row in rows) {
                    val key = row["key"]?.stringValue
                        ?: row["filter_key"]?.stringValue
                        ?: row["service_key"]?.stringValue
                    val action = row["action"]?.intValue
                    if (key != null && action != null) out[key] = action
                }
            } else {
                value.objectValue?.forEach { (key, v) ->
                    (v.intValue ?: v["action"]?.intValue)?.let { out[key] = it }
                }
            }
            return out
        }
    }
}

/**
 * A custom domain rule (`sbr_scheduled_block_rules`). [hardBlock] marks it for
 * the strict-mode tunnel's connection-level enforcement.
 */
data class DnsDomainRule(
    val ruleId: Int,
    val hostname: String,
    /** 0 = block, 1 = allow. */
    val action: Int,
    val isActive: Boolean,
    val hardBlock: Boolean,
) {
    val isBlock: Boolean get() = action == 0

    companion object {
        fun from(json: JsonValue?): DnsDomainRule? {
            val j = json ?: return null
            val ruleId = j["rule_id"]?.intValue ?: return null
            return DnsDomainRule(
                ruleId = ruleId,
                hostname = j["hostname"]?.stringValue ?: "",
                action = j["action"]?.intValue ?: 0,
                isActive = j["is_active"]?.boolValue ?: true,
                hardBlock = j["hard_block"]?.boolValue ?: false,
            )
        }
    }
}

// MARK: - account_summary

/**
 * The five ScrollDaddy feature flags plus device count vs. limit. The client
 * renders locked/upsell states from these; the server rejects gated writes
 * regardless (tier gating is server-enforced).
 */
data class DnsAccountSummary(
    val tierName: String?,
    val maxDevices: Int,
    val maxScheduledBlocks: Int,
    val customRules: Boolean,
    val advancedFilters: Boolean,
    val queryLogging: Boolean,
    val deviceCount: Int,
    val deviceMax: Int,
) {
    val atDeviceLimit: Boolean get() = deviceCount >= deviceMax

    companion object {
        fun from(data: JsonValue?): DnsAccountSummary? {
            if (data == null) return null
            val features = data["features"]
            val maxDevices = features?.get("scrolldaddy_max_devices")?.intValue ?: 1
            return DnsAccountSummary(
                tierName = data["tier_name"]?.stringValue,
                maxDevices = maxDevices,
                maxScheduledBlocks = features?.get("scrolldaddy_max_scheduled_blocks")?.intValue ?: 0,
                customRules = features?.get("scrolldaddy_custom_rules")?.boolValue ?: false,
                advancedFilters = features?.get("scrolldaddy_advanced_filters")?.boolValue ?: false,
                queryLogging = features?.get("scrolldaddy_query_logging")?.boolValue ?: false,
                deviceCount = data["device_count"]?.intValue ?: 0,
                deviceMax = data["device_max"]?.intValue ?: maxDevices,
            )
        }
    }
}

// MARK: - catalog

/**
 * The filter/service catalog, static per deployment (cache client-side). The
 * `advanced` flag gates a filter behind `scrolldaddy_advanced_filters`.
 */
data class DnsCatalog(
    val filters: List<DnsCatalogFilter>,
    val serviceCategories: List<DnsCatalogServiceCategory>,
    /** category_key -> the services in that category. */
    val services: Map<String, List<DnsCatalogService>>,
) {
    /** General (ungated) filters — the free-tier floor, in server order. */
    val generalFilters: List<DnsCatalogFilter> get() = filters.filter { !it.advanced }
    val advancedFiltersList: List<DnsCatalogFilter> get() = filters.filter { it.advanced }

    companion object {
        fun from(data: JsonValue?): DnsCatalog? {
            if (data == null) return null
            val svc = LinkedHashMap<String, List<DnsCatalogService>>()
            data["services"]?.objectValue?.forEach { (key, value) ->
                svc[key] = (value.arrayValue ?: emptyList()).mapNotNull { DnsCatalogService.from(it) }
            }
            return DnsCatalog(
                filters = (data["filters"]?.arrayValue ?: emptyList()).mapNotNull { DnsCatalogFilter.from(it) },
                serviceCategories = (data["service_categories"]?.arrayValue ?: emptyList())
                    .mapNotNull { DnsCatalogServiceCategory.from(it) },
                services = svc,
            )
        }
    }
}

data class DnsCatalogFilter(val key: String, val label: String, val advanced: Boolean) {
    companion object {
        fun from(json: JsonValue?): DnsCatalogFilter? {
            val j = json ?: return null
            val key = j["key"]?.stringValue ?: return null
            return DnsCatalogFilter(key, j["label"]?.stringValue ?: key, j["advanced"]?.boolValue ?: false)
        }
    }
}

data class DnsCatalogServiceCategory(val key: String, val label: String) {
    companion object {
        fun from(json: JsonValue?): DnsCatalogServiceCategory? {
            val j = json ?: return null
            val key = j["key"]?.stringValue ?: return null
            return DnsCatalogServiceCategory(key, j["label"]?.stringValue ?: key)
        }
    }
}

data class DnsCatalogService(val key: String, val label: String) {
    companion object {
        fun from(json: JsonValue?): DnsCatalogService? {
            val j = json ?: return null
            val key = j["key"]?.stringValue ?: return null
            return DnsCatalogService(key, j["label"]?.stringValue ?: key)
        }
    }
}

// MARK: - protection mode

/**
 * The one "protection level" control the app presents. Standard and Strict are
 * a routing-scope switch inside one VpnService (unlike iOS's two mechanisms),
 * so the app owns the switch between them.
 */
enum class ProtectionMode(val slug: String) {
    /** DoH-forwarding tunnel that claims only DNS — policy stays server-side. */
    STANDARD("standard"),

    /** All-traffic tunnel that additionally drops connections by SNI/IP from the
     *  hard-block list. */
    STRICT("strict");

    val title: String
        get() = when (this) {
            STANDARD -> "Standard"
            STRICT -> "Strict"
        }

    val summary: String
        get() = when (this) {
            STANDARD ->
                "Encrypted DNS filtering. Blocks sites at lookup — minimal battery."
            STRICT ->
                "Adds connection-level hard-blocking for selected sites, even when an app brings its own DNS. Traffic never leaves your phone."
        }
}

File: ios/joinery-kit/Sources/JoineryDNSFilterKit/DNSFilterModels.swift
import Foundation
import JoineryKit

// Typed models over the `dns_filtering/` API surface
// (plugins/dns_filtering/docs/overview.md § API Surface). Every shape here is
// the JSON-clean export the server emits (ScrollDaddyHelper::exportDevice /
// exportBlock, account_summary, catalog) — no brand knowledge, so a second
// ScrollDaddy-style deployment (NetworkSentry) parses identically.

// MARK: - devices

/// One device from the `devices` action. `dohURL` is what the standard-mode
/// activation flow saves into `NEDNSSettingsManager`; `hardBlockHostnames` is
/// what strict mode's packet tunnel enforces at the connection level.
public struct DNSDevice: Identifiable, Equatable, Sendable {
    public let deviceID: Int
    public let name: String
    public let deviceName: String
    public let deviceType: String
    public let timezone: String
    public let isActive: Bool
    public let logQueries: Bool
    public let filtersEditable: Bool
    public let resolverUID: String
    public let dohURL: String
    public let dotHostname: String
    public let hardBlockHostnames: [String]
    public let lastSeen: String?
    public let blocks: [DNSBlockSummary]

    public var id: Int { deviceID }

    /// The always-on block is the baseline policy every device carries.
    public var alwaysOnBlock: DNSBlockSummary? {
        blocks.first(where: { $0.isAlwaysOn })
    }

    public var scheduledBlocks: [DNSBlockSummary] {
        blocks.filter { !$0.isAlwaysOn }
    }

    public init?(json: JSONValue?) {
        guard let json, let deviceID = json["device_id"]?.intValue else { return nil }
        self.deviceID = deviceID
        name = json["name"]?.stringValue ?? ""
        deviceName = json["device_name"]?.stringValue ?? ""
        deviceType = json["device_type"]?.stringValue ?? ""
        timezone = json["timezone"]?.stringValue ?? "UTC"
        isActive = json["is_active"]?.boolValue ?? true
        logQueries = json["log_queries"]?.boolValue ?? false
        filtersEditable = json["filters_editable"]?.boolValue ?? true
        resolverUID = json["resolver_uid"]?.stringValue ?? ""
        dohURL = json["doh_url"]?.stringValue ?? ""
        dotHostname = json["dot_hostname"]?.stringValue ?? ""
        hardBlockHostnames = (json["hard_block_hostnames"]?.arrayValue ?? []).compactMap(\.stringValue)
        // `last_seen` arrives as an object ({seen: ...}) proxied from the DNS
        // server, as a bare string, or null. Fold all three to the timestamp.
        lastSeen = json["last_seen"]?["seen"]?.stringValue ?? json["last_seen"]?.stringValue
        blocks = (json["blocks"]?.arrayValue ?? []).compactMap(DNSBlockSummary.init(json:))
    }
}

/// The `blocks` summary rows on a device (no rule contents — `rule_count`
/// only). The full contents come from `block_list` / `scheduled_block_edit`.
public struct DNSBlockSummary: Identifiable, Equatable, Sendable {
    public let blockID: Int
    public let name: String
    public let isAlwaysOn: Bool
    public let isActive: Bool
    public let activeNow: Bool
    public let ruleCount: Int
    public let schedule: DNSSchedule

    public var id: Int { blockID }

    public init?(json: JSONValue?) {
        guard let json, let blockID = json["block_id"]?.intValue else { return nil }
        self.blockID = blockID
        name = json["name"]?.stringValue ?? ""
        isAlwaysOn = json["is_always_on"]?.boolValue ?? false
        isActive = json["is_active"]?.boolValue ?? true
        activeNow = json["active_now"]?.boolValue ?? false
        ruleCount = json["rule_count"]?.intValue ?? 0
        schedule = DNSSchedule(json: json["schedule"])
    }
}

public struct DNSSchedule: Equatable, Sendable {
    public let start: String?
    public let end: String?
    public let days: [Int]
    public let timezone: String?

    public init(json: JSONValue?) {
        start = json?["start"]?.stringValue
        end = json?["end"]?.stringValue
        days = (json?["days"]?.arrayValue ?? []).compactMap(\.intValue)
        timezone = json?["timezone"]?.stringValue
    }
}

// MARK: - block contents (scheduled_block_edit read / block_list)

/// Full contents of one block: category filters, service toggles, and custom
/// domain rules. The always-on editor loads this to render its Block/Allow
/// state.
public struct DNSBlockContents: Equatable, Sendable {
    public let blockID: Int
    public let name: String
    public let isAlwaysOn: Bool
    public let schedule: DNSSchedule
    /// filter_key -> action (0 block, 1 allow). Absent key == Allow ("no row").
    public let filters: [String: Int]
    public let services: [String: Int]
    public let rules: [DNSDomainRule]

    public init?(data: JSONValue?) {
        guard let data else { return nil }
        // scheduled_block_edit read wraps the block under `block`; be lenient
        // and accept a flat shape too.
        let block = data["block"] ?? data
        guard let blockID = block["block_id"]?.intValue else { return nil }
        self.blockID = blockID
        name = block["name"]?.stringValue ?? ""
        isAlwaysOn = block["is_always_on"]?.boolValue ?? false
        schedule = DNSSchedule(json: block["schedule"])

        filters = Self.actionMap(block["filters"])
        services = Self.actionMap(block["services"])
        rules = (block["rules"]?.arrayValue ?? []).compactMap(DNSDomainRule.init(json:))
    }

    /// The block's filter/service rows arrive either as a list of
    /// `{key/filter_key, action}` objects or as a `{key: action}` map,
    /// depending on the action. Fold both into `[key: action]`.
    private static func actionMap(_ value: JSONValue?) -> [String: Int] {
        guard let value else { return [:] }
        var out: [String: Int] = [:]
        if let rows = value.arrayValue {
            for row in rows {
                let key = row["key"]?.stringValue
                    ?? row["filter_key"]?.stringValue
                    ?? row["service_key"]?.stringValue
                if let key, let action = row["action"]?.intValue {
                    out[key] = action
                }
            }
        } else if let pairs = value.objectValue {
            for (key, v) in pairs {
                if let action = v.intValue { out[key] = action }
                else if let action = v["action"]?.intValue { out[key] = action }
            }
        }
        return out
    }
}

/// A custom domain rule (`sbr_scheduled_block_rules`). `hardBlock` marks it
/// for the strict-mode tunnel's connection-level enforcement.
public struct DNSDomainRule: Identifiable, Equatable, Sendable {
    public let ruleID: Int
    public let hostname: String
    /// 0 = block, 1 = allow.
    public let action: Int
    public let isActive: Bool
    public let hardBlock: Bool

    public var id: Int { ruleID }
    public var isBlock: Bool { action == 0 }

    public init?(json: JSONValue?) {
        guard let json, let ruleID = json["rule_id"]?.intValue else { return nil }
        self.ruleID = ruleID
        hostname = json["hostname"]?.stringValue ?? ""
        action = json["action"]?.intValue ?? 0
        isActive = json["is_active"]?.boolValue ?? true
        hardBlock = json["hard_block"]?.boolValue ?? false
    }
}

// MARK: - account_summary

/// The five ScrollDaddy feature flags plus device count vs. limit. The client
/// renders locked/upsell states from these; the server rejects gated writes
/// regardless (tier gating is server-enforced).
public struct DNSAccountSummary: Equatable, Sendable {
    public let tierName: String?
    public let maxDevices: Int
    public let maxScheduledBlocks: Int
    public let customRules: Bool
    public let advancedFilters: Bool
    public let queryLogging: Bool
    public let deviceCount: Int
    public let deviceMax: Int

    public var atDeviceLimit: Bool { deviceCount >= deviceMax }

    public init?(data: JSONValue?) {
        guard let data else { return nil }
        tierName = data["tier_name"]?.stringValue
        let features = data["features"]
        maxDevices = features?["scrolldaddy_max_devices"]?.intValue ?? 1
        maxScheduledBlocks = features?["scrolldaddy_max_scheduled_blocks"]?.intValue ?? 0
        customRules = features?["scrolldaddy_custom_rules"]?.boolValue ?? false
        advancedFilters = features?["scrolldaddy_advanced_filters"]?.boolValue ?? false
        queryLogging = features?["scrolldaddy_query_logging"]?.boolValue ?? false
        deviceCount = data["device_count"]?.intValue ?? 0
        deviceMax = data["device_max"]?.intValue ?? maxDevices
    }
}

// MARK: - catalog

/// The filter/service catalog, static per deployment (cache client-side). The
/// `advanced` flag gates a filter behind `scrolldaddy_advanced_filters`.
public struct DNSCatalog: Equatable, Sendable {
    public let filters: [DNSCatalogFilter]
    public let serviceCategories: [DNSCatalogServiceCategory]
    /// category_key -> the services in that category.
    public let services: [String: [DNSCatalogService]]

    public init?(data: JSONValue?) {
        guard let data else { return nil }
        filters = (data["filters"]?.arrayValue ?? []).compactMap(DNSCatalogFilter.init(json:))
        serviceCategories = (data["service_categories"]?.arrayValue ?? []).compactMap(DNSCatalogServiceCategory.init(json:))
        var svc: [String: [DNSCatalogService]] = [:]
        for (key, value) in (data["services"]?.objectValue ?? []) {
            svc[key] = (value.arrayValue ?? []).compactMap(DNSCatalogService.init(json:))
        }
        services = svc
    }

    /// General (ungated) filters — the free-tier floor. Ordered as the server
    /// returned them.
    public var generalFilters: [DNSCatalogFilter] { filters.filter { !$0.advanced } }
    public var advancedFilters: [DNSCatalogFilter] { filters.filter { $0.advanced } }
}

public struct DNSCatalogFilter: Identifiable, Equatable, Sendable {
    public let key: String
    public let label: String
    public let advanced: Bool
    public var id: String { key }

    public init?(json: JSONValue?) {
        guard let json, let key = json["key"]?.stringValue else { return nil }
        self.key = key
        label = json["label"]?.stringValue ?? key
        advanced = json["advanced"]?.boolValue ?? false
    }
}

public struct DNSCatalogServiceCategory: Identifiable, Equatable, Sendable {
    public let key: String
    public let label: String
    public var id: String { key }

    public init?(json: JSONValue?) {
        guard let json, let key = json["key"]?.stringValue else { return nil }
        self.key = key
        label = json["label"]?.stringValue ?? key
    }
}

public struct DNSCatalogService: Identifiable, Equatable, Sendable {
    public let key: String
    public let label: String
    public var id: String { key }

    public init?(json: JSONValue?) {
        guard let json, let key = json["key"]?.stringValue else { return nil }
        self.key = key
        label = json["label"]?.stringValue ?? key
    }
}

// MARK: - protection mode

/// The one "protection level" control the app presents. Standard and Strict
/// are mutually exclusive at the OS level (an active VPN supersedes installed
/// DNS settings), so the app owns the switch between them.
public enum ProtectionMode: String, Sendable, CaseIterable, Identifiable {
    /// Encrypted DNS via `NEDNSSettingsManager` — policy stays server-side.
    case standard
    /// Local `NEPacketTunnelProvider` adding connection-level hard blocking.
    case strict

    public var id: String { rawValue }

    public var title: String {
        switch self {
        case .standard: return "Standard"
        case .strict: return "Strict"
        }
    }

    public var summary: String {
        switch self {
        case .standard:
            return "Encrypted DNS filtering. Blocks sites at lookup — no VPN, minimal battery."
        case .strict:
            return "Adds an on-device VPN that hard-blocks selected sites at the connection level, even when an app brings its own DNS. Traffic never leaves your phone."
        }
    }
}
